lean4-infoview/src/infoview/infos.tsx

```diff
@@ -29,19 +29,38 @@ export function Infos() {
                         changed = true;
                         return null;
                     }
-                    if (!PositionHelpers.isAfterOrEqual(pin, chg.range.start)) continue;
-
-                    let lines = 0;
-                    for (const c of chg.text) if (c === '\n') lines++;
-                    newPin.line = chg.range.start.line + Math.max(0, newPin.line - chg.range.end.line) + lines;
-                    newPin.character = newPin.line > chg.range.end.line ?
-                        newPin.character :
-                        lines === 0 ?
-                            chg.range.start.character + Math.max(0, newPin.character - chg.range.end.character) + chg.text.length :
-                            9999;
+                    if (PositionHelpers.isLessThanOrEqual(newPin, chg.range.start)) continue;
+                    // We can assume chg.range.start < pin
+
+                    // If the pinned position is replaced with new text, just delete the pin.
+                    if (PositionHelpers.isLessThanOrEqual(newPin, chg.range.end)) {
```

Comment:
is LessThanOrEqual correct here?  If I pin the end of this line:
```lean
#eval [1,2,3]
```
Then replace the string `[1,2,3]` with `456` the pin disappears which is not really expected in this case?

Reply:
Yep, the start/end of the line are edge cases that have to be handled somehow, so I chose to do it this way. The interpretation in this PR is that the pin is "at" the character preceding the cursor when the position is pinned. In your example, when the cursor is at the end, the pin is at `]`. When replaced with `456`, the bracket is gone and so is the pin. The symmetric choice would be to go with the character *after* the cursor. But that would be a newline which is just weird (it seems nonsensical to pin a newline).